File: android/core/protocol/src/main/kotlin/fr/mina/gateway/protocol/AssistantResponseStream.kt
package fr.mina.gateway.protocol

import org.json.JSONObject
import java.nio.ByteBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets

data class AssistantResponseFrame(
    val type: String,
    val responseId: String,
    val sourceEventId: String,
    val sequence: Int,
    val text: String?,
    val code: String?,
)

/**
 * Contrat des payloads v2 de réponse progressive — miroir de
 * `src/contracts/assistant-response-stream.mjs` côté PC.
 */
object AssistantResponseStream {
    const val MAX_SEQUENCE = 999
    const val MAX_CHUNK_BYTES = 8 * 1024
    const val MAX_FINAL_BYTES = 32 * 1024

    val types = setOf(
        "assistant.response.started",
        "assistant.response.chunk",
        "assistant.response.completed",
        "assistant.response.failed",
    )
    val failedCodes = setOf(
        "generation_cancelled",
        "generation_failed",
        "provider_timeout",
        "provider_unavailable",
    )

    private val ulidPattern = Regex("^[0-9A-HJKMNP-TV-Z]{26}$")

    fun encodeStarted(
        responseId: String,
        sourceEventId: String,
        text: String? = null,
        code: String? = null,
    ): ByteArray = encode("assistant.response.started", responseId, sourceEventId, 0, text, code)

    fun encodeChunk(responseId: String, sourceEventId: String, sequence: Int, text: String): ByteArray =
        encode("assistant.response.chunk", responseId, sourceEventId, sequence, text, null)

    fun encodeCompleted(responseId: String, sourceEventId: String, sequence: Int, text: String): ByteArray =
        encode("assistant.response.completed", responseId, sourceEventId, sequence, text, null)

    fun encodeFailed(responseId: String, sourceEventId: String, sequence: Int, code: String): ByteArray =
        encode("assistant.response.failed", responseId, sourceEventId, sequence, null, code)

    fun decode(payload: ChatPayloadCodec.PayloadV2): AssistantResponseFrame {
        require(payload.type in types) { "assistant_response_payload_invalid" }
        val meta = decodeMeta(payload.type, payload.metaJson)
        validateFrame(payload.type, meta.responseId, meta.sourceEventId, meta.sequence, payload.binary, meta.code)
        val text = if (payload.binary.isEmpty()) null else decodeUtf8(payload.binary, bodyError(payload.type))
        return AssistantResponseFrame(payload.type, meta.responseId, meta.sourceEventId, meta.sequence, text, meta.code)
    }

    private fun encode(
        type: String,
        responseId: String,
        sourceEventId: String,
        sequence: Int,
        text: String?,
        code: String?,
    ): ByteArray {
        require(type in types) { "assistant_response_type_invalid" }
        val body = encodeBody(type, text)
        validateFrame(type, responseId, sourceEventId, sequence, body, code)
        val meta = JSONObject()
            .put("responseId", responseId)
            .put("sourceEventId", sourceEventId)
            .put("sequence", sequence)
        if (code != null) meta.put("code", code)
        return ChatPayloadCodec.encodeV2(type, meta.toString(), body)
    }

    private fun encodeBody(type: String, text: String?): ByteArray {
        if (type == "assistant.response.started" || type == "assistant.response.failed") {
            require(text == null) { bodyError(type) }
            return ByteArray(0)
        }
        return encodeUtf8(text, bodyError(type))
    }

    private fun encodeUtf8(text: String?, error: String): ByteArray {
        require(text != null) { error }
        val bytes = text.toByteArray(StandardCharsets.UTF_8)
        require(decodeUtf8(bytes, error) == text) { error }
        return bytes
    }

    private fun decodeUtf8(bytes: ByteArray, error: String): String = try {
        StandardCharsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
            .decode(ByteBuffer.wrap(bytes))
            .toString()
    } catch (_: CharacterCodingException) {
        throw IllegalArgumentException(error)
    }

    private fun validateFrame(
        type: String,
        responseId: String,
        sourceEventId: String,
        sequence: Int,
        body: ByteArray,
        code: String?,
    ) {
        require(type in types) { "assistant_response_type_invalid" }
        require(ulidPattern.matches(responseId) && ulidPattern.matches(sourceEventId)) { "assistant_response_id_invalid" }
        require(sequence in 0..MAX_SEQUENCE) { "assistant_response_sequence_invalid" }

        when (type) {
            "assistant.response.started" ->
                require(sequence == 0 && body.isEmpty() && code == null) { "assistant_response_started_body_invalid" }
            "assistant.response.chunk" -> {
                require(sequence >= 1) { "assistant_response_sequence_invalid" }
                require(body.isNotEmpty() && body.size <= MAX_CHUNK_BYTES && code == null) {
                    "assistant_response_chunk_body_invalid"
                }
            }
            "assistant.response.completed" -> {
                require(sequence >= 1) { "assistant_response_sequence_invalid" }
                require(body.isNotEmpty() && body.size <= MAX_FINAL_BYTES && code == null) {
                    "assistant_response_completed_body_invalid"
                }
            }
            "assistant.response.failed" -> {
                require(sequence >= 1) { "assistant_response_sequence_invalid" }
                require(body.isEmpty()) { "assistant_response_failed_body_invalid" }
                require(code in failedCodes) { "assistant_response_code_invalid" }
            }
        }
    }

    private data class ResponseMeta(
        val responseId: String,
        val sourceEventId: String,
        val sequence: Int,
        val code: String?,
    )

    private fun decodeMeta(type: String, metaJson: String): ResponseMeta {
        val json = try {
            JSONObject(metaJson)
        } catch (_: Exception) {
            throw IllegalArgumentException("assistant_response_meta_invalid")
        }
        val expected = if (type == "assistant.response.failed") {
            setOf("responseId", "sourceEventId", "sequence", "code")
        } else {
            setOf("responseId", "sourceEventId", "sequence")
        }
        require(json.keys().asSequence().toSet() == expected) { "assistant_response_meta_invalid" }
        val responseId = json.opt("responseId") as? String ?: throw IllegalArgumentException("assistant_response_meta_invalid")
        val sourceEventId = json.opt("sourceEventId") as? String ?: throw IllegalArgumentException("assistant_response_meta_invalid")
        val sequence = json.opt("sequence") as? Int ?: throw IllegalArgumentException("assistant_response_meta_invalid")
        val code = if (type == "assistant.response.failed") {
            json.opt("code") as? String ?: throw IllegalArgumentException("assistant_response_meta_invalid")
        } else {
            null
        }
        return ResponseMeta(responseId, sourceEventId, sequence, code)
    }

    private fun bodyError(type: String): String = "assistant_response_${type.substringAfterLast('.')}_body_invalid"
}

File: src/contracts/assistant-response-stream.mjs
import { TextDecoder } from 'node:util';
import { decodeChatPayload, encodeChatPayloadV2 } from './chat-payload.mjs';

export const ASSISTANT_RESPONSE_FRAME_TYPES = Object.freeze([
  'assistant.response.started',
  'assistant.response.chunk',
  'assistant.response.completed',
  'assistant.response.failed',
]);

export const MAX_ASSISTANT_RESPONSE_SEQUENCE = 999;
export const MAX_ASSISTANT_RESPONSE_CHUNK_BYTES = 8 * 1024;
export const MAX_ASSISTANT_RESPONSE_FINAL_BYTES = 32 * 1024;
export const ASSISTANT_RESPONSE_FAILED_CODES = Object.freeze([
  'generation_cancelled',
  'generation_failed',
  'provider_timeout',
  'provider_unavailable',
]);

const RESPONSE_TYPES = new Set(ASSISTANT_RESPONSE_FRAME_TYPES);
const FAILED_CODES = new Set(ASSISTANT_RESPONSE_FAILED_CODES);
const ULID_PATTERN = /^[0-9A-HJKMNP-TV-Z]{26}$/u;
const UTF8_DECODER = new TextDecoder('utf-8', { fatal: true, ignoreBOM: true });

function fail(code) {
  throw new Error(code);
}

function bodyError(type) {
  return `assistant_response_${type.split('.').at(-1)}_body_invalid`;
}

function decodeUtf8(binary, code) {
  try {
    return UTF8_DECODER.decode(binary);
  } catch {
    fail(code);
  }
}

function encodeUtf8(text, code) {
  if (typeof text !== 'string') fail(code);
  const binary = Buffer.from(text, 'utf8');
  if (decodeUtf8(binary, code) !== text) fail(code);
  return binary;
}

function assertMeta(type, meta) {
  if (meta === null || Array.isArray(meta) || typeof meta !== 'object') fail('assistant_response_meta_invalid');
  const expected = type === 'assistant.response.failed'
    ? ['code', 'responseId', 'sequence', 'sourceEventId']
    : ['responseId', 'sequence', 'sourceEventId'];
  const keys = Object.keys(meta).sort();
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    fail('assistant_response_meta_invalid');
  }
  if (typeof meta.responseId !== 'string' || typeof meta.sourceEventId !== 'string' || !Number.isInteger(meta.sequence)) {
    fail('assistant_response_meta_invalid');
  }
  if (type === 'assistant.response.failed' && typeof meta.code !== 'string') {
    fail('assistant_response_meta_invalid');
  }
  return {
    responseId: meta.responseId,
    sourceEventId: meta.sourceEventId,
    sequence: meta.sequence,
    code: type === 'assistant.response.failed' ? meta.code : null,
  };
}

function validateFrame({ type, responseId, sourceEventId, sequence, body, code }) {
  if (!RESPONSE_TYPES.has(type)) fail('assistant_response_type_invalid');
  if (!ULID_PATTERN.test(responseId) || !ULID_PATTERN.test(sourceEventId)) fail('assistant_response_id_invalid');
  if (!Number.isInteger(sequence) || sequence < 0 || sequence > MAX_ASSISTANT_RESPONSE_SEQUENCE) {
    fail('assistant_response_sequence_invalid');
  }

  if (type === 'assistant.response.started') {
    if (sequence !== 0 || body.length !== 0 || code !== null) fail('assistant_response_started_body_invalid');
    return;
  }

  if (sequence < 1) fail('assistant_response_sequence_invalid');
  if (type === 'assistant.response.chunk') {
    if (body.length === 0 || body.length > MAX_ASSISTANT_RESPONSE_CHUNK_BYTES || code !== null) {
      fail('assistant_response_chunk_body_invalid');
    }
    return;
  }
  if (type === 'assistant.response.completed') {
    if (body.length === 0 || body.length > MAX_ASSISTANT_RESPONSE_FINAL_BYTES || code !== null) {
      fail('assistant_response_completed_body_invalid');
    }
    return;
  }
  if (body.length !== 0) fail('assistant_response_failed_body_invalid');
  if (!FAILED_CODES.has(code)) fail('assistant_response_code_invalid');
}

function encodeBody(type, text) {
  if (type === 'assistant.response.started' || type === 'assistant.response.failed') {
    if (text !== undefined && text !== null) fail(bodyError(type));
    return Buffer.alloc(0);
  }
  return encodeUtf8(text, bodyError(type));
}

/**
 * Encode une trame de réponse dans un payload v2 avant chiffrement de l'enveloppe mina_app.
 * Le corps final/chunk reste dans la section binaire pour séparer explicitement le texte
 * sensible des métadonnées de corrélation.
 */
export function encodeAssistantResponseFrame({ type, responseId, sourceEventId, sequence, text, code = null } = {}) {
  if (!RESPONSE_TYPES.has(type)) fail('assistant_response_type_invalid');
  const body = encodeBody(type, text);
  validateFrame({ type, responseId, sourceEventId, sequence, body, code });
  const meta = { responseId, sourceEventId, sequence };
  if (code !== null) meta.code = code;
  return encodeChatPayloadV2({ type, meta, binary: body });
}

/** Décode et valide une trame authentifiée de réponse progressive. */
export function decodeAssistantResponseFrame(payload) {
  const decoded = decodeChatPayload(payload);
  if (decoded.version !== 2 || !RESPONSE_TYPES.has(decoded.type)) fail('assistant_response_payload_invalid');
  const { responseId, sourceEventId, sequence, code } = assertMeta(decoded.type, decoded.meta);
  const body = Buffer.from(decoded.binary);
  validateFrame({ type: decoded.type, responseId, sourceEventId, sequence, body, code });
  const text = body.length === 0 ? null : decodeUtf8(body, bodyError(decoded.type));
  return Object.freeze({ type: decoded.type, responseId, sourceEventId, sequence, text, code });
}
